src/ejabberd_c2s.erl

```diff
@@ -737,47 +737,46 @@ do_open_session(Acc, JID, StateData) ->
             end
     end.
 
-do_open_session_common(Acc, JID, #state{user = U, resource = R} = NewStateData0) ->
+do_open_session_common(Acc, JID, #state{user = U, server = S, resource = R} = NewStateData0) ->
     change_shaper(NewStateData0, JID),
-    Acc1 = ejabberd_hooks:run_fold(roster_get_subscription_lists,
-                                  NewStateData0#state.server,
-                                  Acc,
-                                  [U, NewStateData0#state.server]),
+    Acc1 = ejabberd_hooks:run_fold(roster_get_subscription_lists, S, Acc, [U, S]),
     {Fs, Ts, Pending} = mongoose_acc:get(roster, subscription_lists, {[], [], []}, Acc1),
     LJID = jid:to_lower(jid:to_bare(JID)),
     Fs1 = [LJID | Fs],
     Ts1 = [LJID | Ts],
-    PrivList = ejabberd_hooks:run_fold(privacy_get_user_list,
-                                       NewStateData0#state.server,
-                                       #userlist{},
-                                       [U, NewStateData0#state.server]),
+    PrivList = ejabberd_hooks:run_fold(privacy_get_user_list, S, #userlist{}, [U, S]),
     SID = {p1_time_compat:timestamp(), self()},
     Conn = get_conn_type(NewStateData0),
     Info = [{ip, NewStateData0#state.ip}, {conn, Conn},
             {auth_module, NewStateData0#state.auth_module}],
-    ReplacedPids = ejabberd_sm:open_session(SID, U, NewStateData0#state.server, R, Info),
+    ReplacedPids = ejabberd_sm:open_session(SID, U, S, R, Info),
 
-    MonitorRefs = ordsets:from_list([{monitor(process, PID), PID} || PID <- ReplacedPids]),
-    lists:foreach(
-        fun({MonitorRef, PID}) ->
-            receive
-                {'DOWN', MonitorRef, _, _, _} -> ok
-            after 100 ->
-                ?WARNING_MSG("C2S process ~p for ~s replaced by ~p has not stopped before timeout",
-                             [PID, jid:to_binary(NewStateData0#state.jid), self()])
-            end
-        end,
-        MonitorRefs),
+    RefsAndPids = [{monitor(process, PID), PID} || PID <- ReplacedPids],
+    case RefsAndPids of
+        [] ->
+            ok;
+        _ ->
+            Timeout = get_replaced_wait_timeout(S),
+            erlang:send_after(Timeout, self(), replaced_wait_timeout)
```

Comment:
I am not sure whether it is crucial, but maybe it would be worth to use `TRef`? In order to prevent c2s process from getting messages `replaced_wait_timeout` from different sources than itself?

Reply:
Is there some another source of such message? Not really :)